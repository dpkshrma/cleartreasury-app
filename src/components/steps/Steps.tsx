import * as React from "react";
import { State } from "./Step";
import Surface from "../surface/Surface";

type StepItemProps = {
  stepTitle?: string;
  form?: React.ReactElement;
  children?: React.ReactElement;
};

type StepsProps = {
  nav?: React.ReactElement;
  children: React.ReactElement[];
  stepBack?(step?: number): void;
  onComplete?(): void;
};

const Step = ({ form, children, ...props }: StepItemProps): JSX.Element => {
  return React.cloneElement(children || form, { ...props });
};

const Steps = ({
  nav,
  children,
  stepBack,
  onComplete,
}: StepsProps): JSX.Element => {
  const [activeStep, setActiveStep] = React.useState(0);
  const [formState, setFormState] = React.useState({});

  React.useEffect(() => {
    if (activeStep < 0) stepBack();
  }, [activeStep]);

  const Nav = (): JSX.Element | null =>
    nav ? (
      <div className="sm:flex justify-center sm:mx-4">
        {children.map((child, index) =>
          React.cloneElement(nav, {
            key: index,
            step: index,
            title: child.props.stepTitle,
            state:
              activeStep === index
                ? State.ACTIVE
                : formState[index] || State.DEFAULT,
            isEnabled: index <= activeStep,
            onClick: () => setActiveStep(index),
          })
        )}
      </div>
    ) : null;

  return (
    <>
      <Nav />

      <Surface align={Surface.Align.CENTER}>
        {children.map((child, index) => {
          return (
            index === activeStep &&
            React.cloneElement(child, {
              key: index,
              stepBack: (step: number) => {
                if (step === activeStep) {
                  stepBack(step);
                } else {
                  isNaN(step)
                    ? setActiveStep(activeStep - 1)
                    : setActiveStep(step);
                }
              },
              onComplete: (data) => {
                child.props.form?.props.onComplete &&
                  child.props.form.props.onComplete(data);

                if (activeStep === children.length - 1) {
                  onComplete();
                } else {
                  setFormState(() => ({
                    ...formState,
                    [activeStep]: State.COMPLETE,
                  }));
                  setActiveStep((activeStep) => activeStep + 1);
                }
              },
            })
          );
        })}
      </Surface>
    </>
  );
};

Steps.Step = Step;

export default Steps;
