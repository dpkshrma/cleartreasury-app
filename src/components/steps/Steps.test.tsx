import * as React from "react";
import { render } from "../../../test/testUtils";
import { StepsStory } from "./Steps.stories";
import { act, cleanup, fireEvent, getByText } from "@testing-library/react";

describe("Steps", () => {
  let component: DocumentFragment;
  let firstChild: HTMLCanvasElement;

  beforeEach(() => {
    const { asFragment } = render(<StepsStory />);
    component = asFragment();
    firstChild = component.firstChild as HTMLCanvasElement;
  });
  afterEach(cleanup);

  it("matches snapshot for nested children", () => {
    expect(component).toMatchSnapshot();
  });

  describe("matches snapshot fragment", () => {
    it("for Amount text", () => {
      const span = getByText(firstChild, "Amount");
      expect(span).not.toBe(null);
      expect(span).toMatchSnapshot();
    });

    it("for Step 1 form", () => {
      const form = firstChild.querySelector("form");

      expect(form).not.toBe(null);
      expect(form).toMatchSnapshot();
    });

    it("for Step 2 form after Step 1 submit", async () => {
      const button = getByText(firstChild, "Continue");
      expect(button).not.toBe(null);

      await act(async () => {
        fireEvent.click(button);
      });

      const beneficiaryForm = firstChild.querySelector("form");
      expect(beneficiaryForm).toMatchSnapshot();
    });
  });
});
