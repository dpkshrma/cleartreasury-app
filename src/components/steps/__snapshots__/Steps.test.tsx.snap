// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps matches snapshot for nested children 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <div
      class="flex justify-between"
    >
      <div
        class="flex flex-1 pb-4 pt-3 px-2 cursor-default border-t-4 border-green-500 bg-white"
      >
        <span
          class="flex justify-center p-0.5 w-6 h-6 rounded-2xl text-xs text-white bg-teal-700"
        >
          1
        </span>
        <span
          class="ml-2"
        >
          Amount
        </span>
      </div>
      <div
        class="flex flex-1 pb-4 pt-3 px-2 cursor-default border-t-4 border-transparent"
      >
        <span
          class="flex justify-center p-0.5 w-6 h-6 rounded-2xl text-xs text-white bg-teal-700"
        >
          2
        </span>
        <span
          class="ml-2"
        >
          Beneficiary
        </span>
      </div>
      <div
        class="flex flex-1 pb-4 pt-3 px-2 cursor-default border-t-4 border-transparent"
      >
        <span
          class="flex justify-center p-0.5 w-6 h-6 rounded-2xl text-xs text-white bg-teal-700"
        >
          3
        </span>
        <span
          class="ml-2"
        >
          Confirm & Pay
        </span>
      </div>
    </div>
    <div>
      <form
        data-testid="form"
      >
        <h1>
          Form 1
        </h1>
        <button>
          Continue
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Steps navigate to next step when submitting a form 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <div
      class="flex justify-between"
    >
      <div
        class="flex flex-1 pb-4 pt-3 px-2 cursor-pointer text-gray-400 border-t-4 border-gray-400"
      >
        <svg
          class="w-7 h-7 flex-shrink-0 font-normal"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.707-9.293a1 1 0 00-1.414-1.414L9 10.586 7.707 9.293a1 1 0 00-1.414 1.414l2 2a1 1 0 001.414 0l4-4z"
            fill-rule="evenodd"
          />
        </svg>
        <span
          class="ml-2"
        >
          Amount
        </span>
      </div>
      <div
        class="flex flex-1 pb-4 pt-3 px-2 cursor-default border-t-4 border-green-500 bg-white"
      >
        <span
          class="flex justify-center p-0.5 w-6 h-6 rounded-2xl text-xs text-white bg-teal-700"
        >
          2
        </span>
        <span
          class="ml-2"
        >
          Beneficiary
        </span>
      </div>
      <div
        class="flex flex-1 pb-4 pt-3 px-2 cursor-default border-t-4 border-transparent"
      >
        <span
          class="flex justify-center p-0.5 w-6 h-6 rounded-2xl text-xs text-white bg-teal-700"
        >
          3
        </span>
        <span
          class="ml-2"
        >
          Confirm & Pay
        </span>
      </div>
    </div>
    <div>
      <div
        class="w-full"
      >
        <div>
          <form
            data-testid="form"
          >
            <h1>
              Form 2A
            </h1>
            <button>
              Continue
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
